new metrics + new vignette: making new layers via subclassing (#1261)
rstudio/keras: vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd

  }
}
```

You would use a layer by calling it on some tensor input(s), much like a regular
function.

```{r}
x <- tf$ones(shape(2, 2))
linear_layer <- Linear(4, 2)
y <- linear_layer(x)
print(y)
```

`Linear` behaves similarly to a layer present in the python interface to keras
(e.g., `keras$layers$Dense`).

However, one additional step is needed to make it behave like the builtin layers
present in the keras R package (e.g., `layer_dense()`).

Keras layers in R are designed to compose nicely with the pipe operator
(`%>%`), so that the layer instance is conveniently created on demand when an
existing model or tensor is piped in. In order to make a custom layer similarly
compose nicely with the pipe, you can call `create_layer_wrapper()` on the layer
class constructor.

```{r}
layer_linear <- create_layer_wrapper(Linear)
```

Now `layer_linear` is a layer constructor that composes nicely with `%>%`, just like the built-in layers:

```{r}
model <- keras_model_sequential() %>%
  layer_linear(4, 2)

model(k_ones(c(2, 2)))

model
```

Because the pattern above is so common, there is a convenience function that
combines the steps of subclassing `keras$layers$Layer` and calling

> `create_layer_wrapper` on the output: the `Layer` function. The `layer_linear`

Reviewer: I like it!

defined below is identical to the `layer_linear` defined above.

```{r}
layer_linear <- Layer(
  "Linear",
  initialize =  function(units = 32, input_dim = 32) {
    super$initialize()
    w_init <- tf$random_normal_initializer()
    self$w <- tf$Variable(initial_value = w_init(shape = shape(input_dim, units),
                                                 dtype = "float32"),
                          trainable = TRUE)
    b_init <- tf$zeros_initializer()
    self$b <- tf$Variable(initial_value = b_init(shape = shape(units),
                                                 dtype = "float32"),
                          trainable = TRUE)
  },

  call = function(inputs) {
    tf$matmul(inputs, self$w) + self$b
  }
)
```

> For the remainder of this vignette we'll be using the `%py_class%` constructor,

Reviewer: not sure if these reasons will be convincing to readers ... I guess I'd rather remove that :-)

encoder <- keras_model(inputs = original_inputs, outputs = z, name = "encoder")

# Define decoder model.
latent_inputs <- layer_input(shape = latent_dim, name = "z_sampling")
x <- layer_dense(units = intermediate_dim, activation = "relu")(latent_inputs)
outputs <- layer_dense(units = original_dim, activation = "sigmoid")(x)
decoder <- keras_model(inputs = latent_inputs, outputs = outputs, name = "decoder")

# Define VAE model.
outputs <- decoder(z)
vae <- keras_model(inputs = original_inputs, outputs = outputs, name = "vae")

# Add KL divergence regularization loss.
kl_loss <- -0.5 * tf$reduce_mean(z_log_var - tf$square(z_mean) - tf$exp(z_log_var) + 1)
vae$add_loss(kl_loss)

# Train.
optimizer <- keras$optimizers$Adam(learning_rate = 1e-3)
vae %>% compile(optimizer, loss = loss_mean_squared_error())
vae %>% fit(x_train, x_train, epochs = 3, batch_size = 64)
```

For more information, make sure to read the [Functional API guide](/guides/functional_api/).

## Defining custom layers and models in an R package

Unfortunately you can't use anything that creates references to python objects (e.g., `%py_class%`), at the top-level of an R package.

> When you build an R package, all the R files in the `R/` directory get sourced

Reviewer: great explanation!

library(tfdatasets)
library(keras)

tf_version()
```

## The `Layer` class: the combination of state (weights) and some computation

One of the central abstraction in Keras is the `Layer` class. A layer
encapsulates both a state (the layer's "weights") and a transformation from
inputs to outputs (a "call", the layer's forward pass).

Here's a densely-connected layer. It has a state: the variables `w` and `b`.
```{r}
Linear(keras$layers$Layer) %py_class% {
  initialize <- function(units = 32, input_dim = 32) {
    super$initialize()
    w_init <- tf$random_normal_initializer()
    self$w <- tf$Variable(
      initial_value = w_init(
        shape = shape(input_dim, units),
        dtype = "float32"
      ),
      trainable = TRUE
    )
    b_init <- tf$zeros_initializer()
    self$b <- tf$Variable(
      initial_value = b_init(shape = shape(units), dtype = "float32"),
      trainable = TRUE
    )
  }

  call <- function(inputs) {
    tf$matmul(inputs, self$w) + self$b
  }
}
```

You would use a layer by calling it on some tensor input(s), much like a regular
function.

```{r}
x <- tf$ones(shape(2, 2))
linear_layer <- Linear(4, 2)
y <- linear_layer(x)
print(y)
```

> All the builtin layers in the keras package (e.g., `layer_dense()`) are

Reviewer: very nicely explained!

Reviewer: on the other hand ... seeing this again some time later, now i think, - won't we always want this?
Do it transparently to the user?

additionally designed to compose nicely with the pipe operator (`%>%`), so that
the layer instance is conveniently created on demand when an existing model or
tensor is piped in. In order to make a custom layer similarly compose nicely
with the pipe, you can call `create_layer_wrapper()`.

```{r}
layer_linear <- create_layer_wrapper(Linear)
layer_linear
# function (object, units = 32, input_dim = 32)
# {
#     args <- capture_args(match.call(), modifiers)
#     create_layer(LayerClass, object, args)
# }
# <environment: 0x557cc3d71158>
```

> Now `layer_linear` is a layer constructor composes nicely with `%>%`, just like the built in layers:

Reviewer: ```suggestion
Now `layer_linear` as a layer constructor composes nicely with `%>%`, just like the built-in layers:
```

```{r}
model <- keras_model_sequential() %>%
  layer_linear(4, 2)

> model(k_ones(c(2, 2))) |> invisible() # force model to build

Reviewer: |> invisible() # force model to build

is this for education purposes? :-))))

In that case, I get it, - but doesn't it still add complexity where none is needed?

model
# Model
# Model: "sequential"
# __________________________________________________________________________________________
# Layer (type)                            Output Shape                        Param #
# ==========================================================================================
# linear_1 (Linear)                       (2, 4)                              12
# ==========================================================================================
# Total params: 12
# Trainable params: 12
# Non-trainable params: 0
# __________________________________________________________________________________________
```

Note that the weights `w` and `b` are automatically tracked by the layer upon
being set as layer attributes:

```{r}
stopifnot(all.equal(
  linear_layer$weights,
  list(linear_layer$w, linear_layer$b)
))
```

> Note you also have access to a quicker shortcut for adding weight to a layer:

Reviewer: ```suggestion
Note you also have access to a quicker shortcut for adding a weight to a layer:
```

> the `add_weight()` method:

Reviewer: i think that if the guide wasn't a translation, i'd only show this way of doing it ...

```{r}
Linear(keras$layers$Layer) %py_class% {
  initialize <- function(units = 32, input_dim = 32) {
    super$initialize()
    w_init <- tf$random_normal_initializer()
    self$w <- self$add_weight(
      shape = shape(input_dim, units),
      initializer = "random_normal",
      trainable = TRUE
    )
    self$b <- self$add_weight(
      shape = shape(units),
      initializer = "zeros",
      trainable = TRUE
    )
  }

  call <- function(inputs) {
    tf$matmul(inputs, self$w) + self$b
  }
}

x <- tf$ones(shape(2, 2))
linear_layer <- Linear(4, 2)
y <- linear_layer(x)
print(y)
```

## Layers can have non-trainable weights

Besides trainable weights, you can add non-trainable weights to a layer as
well. Such weights are meant not to be taken into account during
backpropagation, when you are training the layer.

> Here's how to add and use a non-trainable weight:

Reviewer: ok, i guess this answers what i just wrote about the `tf$Variable` way ...

cat("non-trainable weights:", length(my_sum$non_trainable_weights), "\n")

# It's not included in the trainable weights:
cat("trainable_weights:", my_sum$trainable_weights, "\n")
```

## Best practice: deferring weight creation until the shape of the inputs is known

Our `Linear` layer above took an `input_dim `argument that was used to compute
the shape of the weights `w` and `b` in `initialize()`:

```{r}
Linear(keras$layers$Layer) %py_class% {
  initialize <- function(units = 32, input_dim = 32) {
    super$initialize()
    self$w <- self$add_weight(
      shape = shape(input_dim, units),
      initializer = "random_normal",
      trainable = TRUE
    )
    self$b <- self$add_weight(
      shape = shape(units),
      initializer = "zeros",
      trainable = TRUE
    )
  }

  call <- function(inputs) {
    tf$matmul(inputs, self$w) + self$b
  }
}
```

In many cases, you may not know in advance the size of your inputs, and you
would like to lazily create weights when that value becomes known, some time
after instantiating the layer.

In the Keras API, we recommend creating layer weights in the `build(self,
inputs_shape)` method of your layer. Like this:

```{r}

> # Same as below, but using R6.

Reviewer: Will you show both in the end? In this case, I'd start with the Python-ey way, because it's what you mostly use, and then say sth like: "If you use keras in a package, ..." etc

# One benefit of R6: this layer class definition initializes python lazily, so
# it can be safely used in a package.
Linear <- R6::R6Class(
  classname = "Linear",
  inherit = keras$layers$Layer,
  public = list(
    initialize = function(units = 32) {
      super$initialize()
      self$units <- units
    },
    build = function(input_shape) {

>       print(sys.call())

Reviewer: guess this should be removed? :-))

      self$w <- self$add_weight(
        shape = shape(tail(input_shape, 1), self$units),
        initializer = "random_normal",
        trainable = TRUE
      )
      self$b <- self$add_weight(
        shape = shape(self$units),
        initializer = "random_normal",
        trainable = TRUE
      )
    },
    call = function(inputs) {
      tf$matmul(inputs, self$w) + self$b
    }
  )
)
layer_linear <- create_layer_wrapper(Linear)

model <- keras_model_sequential() %>%
  layer_linear(32)
model(k_ones(c(2, 2))) |> invisible() # force model to build
model
```

```{r}
Linear(keras$layers$Layer) %py_class% {
  initialize <- function(units = 32) {
    super$initialize()
    self$units <- units
  }

  build <- function(input_shape) {
    self$w <- self$add_weight(
      shape = shape(tail(input_shape, 1), self$units),
      initializer = "random_normal",
      trainable = TRUE
    )
    self$b <- self$add_weight(
      shape = shape(self$units),
      initializer = "random_normal",
      trainable = TRUE
    )
  }

  call <- function(inputs) {
    tf$matmul(inputs, self$w) + self$b
  }
}
```

The `build()` method of your layer will automatically run the first time

> your layer instance is called. You now have a layer that's lazy and thus easier to use:

Reviewer: ```suggestion
your layer instance is called. You now have a layer that can handle an arbitrary number of input features:
```

    self$linear_1 <- Linear(32)
    self$linear_2 <- Linear(32)
    self$linear_3 <- Linear(1)
  }

  call <- function(inputs) {
    x <- self$linear_1(inputs)
    x <- tf$nn$relu(x)
    x <- self$linear_2(x)
    x <- tf$nn$relu(x)
    self$linear_3(x)
  }
}

mlp <- MLPBlock()
y <- mlp(tf$ones(shape = shape(3, 64))) # The first call to the `mlp` will create the weights
cat("weights:", length(mlp$weights), "\n")
cat("trainable weights:", length(mlp$trainable_weights), "\n")
```

## The `add_loss()` method

When writing the `call()` method of a layer, you can create loss tensors that
you will want to use later, when writing your training loop. This is doable by
calling `self$add_loss(value)`:

```{r}
# A layer that creates an activity regularization loss
ActivityRegularizationLayer(keras$layers$Layer) %py_class% {
  initialize <- function(rate = 1e-2) {
    super$initialize()
    self$rate <- rate
  }

  call <- function(inputs) {
    self$add_loss(self$rate * tf$reduce_sum(inputs))
    inputs
  }
}
# layer_activity_regularization <- create_layer_wrapper(ActivityRegularizationLayer)
```

These losses (including those created by any inner layer) can be retrieved via

> `layer.losses`. This property is reset at the start of every `__call__()` to

Reviewer: ```suggestion
`layer$losses`. This property is reset at the start of every `call()` to
```

# `layer$losses` gets reset at the start of each call()
layer(tf$zeros(shape(1, 1))) |> invisible()
stopifnot(length(layer$losses) == 1) # This is the loss created during the call above
```

In addition, the `loss` property also contains regularization losses created
for the weights of any inner layer:

```{r}
OuterLayerWithKernelRegularizer(keras$layers$Layer) %py_class% {
  initialize <- function() {
    super$initialize()
    self$dense <- layer_dense(units = 32, kernel_regularizer = regularizer_l2(1e-3))
  }
  call <- function(inputs) {
    self$dense(inputs)
  }
}

layer <- OuterLayerWithKernelRegularizer()
layer(tf$zeros(shape(1, 1))) |> invisible()

# This is `1e-3 * sum(layer$dense$kernel ** 2)`,
# created by the `kernel_regularizer` above.
print(layer$losses)
```

These losses are meant to be taken into account when writing training loops,
like this:

```{r}
zip <- function(..., simplify = TRUE) {
  .mapply(if (simplify) c else list, list(...), NULL)
}
```

```{r, eval=FALSE}
# Instantiate an optimizer.
optimizer <- optimizer_sgd(learning_rate = 1e-3)
loss_fn <- loss_sparse_categorical_crossentropy(from_logits = TRUE)

> # while(!is.null(batch <- iter_next(dataset_iterator))) { ... } ## too compact?

Reviewer: i find this easiest - the "while" way i'm referring to

    # Loss value for this minibatch
    loss_value <- loss_fn(y_batch_train, logits)
    # Add extra losses created during this forward pass:
    loss_value <- loss_value + sum(model$losses)
  })
  grads <- tape$gradient(loss_value, model$trainable_weights)
  optimizer$apply_gradients(zip(grads, model$trainable_weights))
}
```

For a detailed guide about writing training loops, see the
[guide to writing a training loop from scratch](/guides/writing_a_training_loop_from_scratch/).

These losses also work seamlessly with `fit()` (they get automatically summed
and added to the main loss, if any):

```{r, eval=TRUE}
input <- layer_input(shape(3))
output <- input %>% layer_activity_regularization()
# output <- ActivityRegularizationLayer()(input)
model <- keras_model(input, output)

# If there is a loss passed in `compile`, the regularization
# losses get added to it
model %>% compile(optimizer = "adam", loss = "mse")
model %>% fit(k_random_uniform(c(2, 3)),
  k_random_uniform(c(2, 3)),
  epochs = 1, verbose = FALSE
)

# It's also possible not to pass any loss in `compile`,
# since the model already has a loss to minimize, via the `add_loss`
# call during the forward pass!

> model %>% compile(optimizer = "adam")

Reviewer: so great this works now! Remember i needed a workaround in tfprobability, in the past...

model %>% fit(k_random_uniform(c(2, 3)),
  k_random_uniform(c(2, 3)),
  epochs = 1, verbose = FALSE
)
```

## The `add_metric()` method

> Similarly to `add_loss()`, layers also have an `add_metric()` method

Reviewer: this is a great addition!

for tracking the moving average of a quantity during training.

Consider the following layer: a "logistic endpoint" layer.
It takes as inputs predictions and targets, it computes a loss which it tracks
via `add_loss()`, and it computes an accuracy scalar, which it tracks via
`add_metric()`.

```{r}
LogisticEndpoint(keras$layers$Layer) %py_class% {
  initialize <- function(name = NULL) {
    super$initialize(name = name)
    # loss_binary_crossentropy(from_logits = TRUE)
    self$loss_fn <- keras$losses$BinaryCrossentropy(from_logits = TRUE)
    self$accuracy_fn <- keras$metrics$BinaryAccuracy()
  }

  call <- function(targets, logits, sample_weights = NULL) {
    # Compute the training-time loss value and add it
    # to the layer using `self$add_loss()`.
    loss <- self$loss_fn(targets, logits, sample_weights)
    self$add_loss(loss)

    # Log accuracy as a metric and add it
    # to the layer using `self.add_metric()`.
    acc <- self$accuracy_fn(targets, logits, sample_weights)
    self$add_metric(acc, name = "accuracy")

    # Return the inference-time prediction tensor (for `.predict()`).
    tf$nn$softmax(logits)
  }
}
```

> Metrics tracked in this way are accessible via `layer.metrics`:

Reviewer: ```suggestion
Metrics tracked in this way are accessible via `layer$metrics`:
```

    config <- super$get_config()
    config$units <- self$units
    config
  }
}

layer <- Linear(64)
config <- layer$get_config()
str(config)
new_layer <- Linear$from_config(config)
```

If you need more flexibility when deserializing the layer from its config, you
can also override the `from_config()` class method. This is the base
implementation of `from_config()`:

```{r}
from_config <- function(cls, config) do.call(cls, config)
```

To learn more about serialization and saving, see the complete
[guide to saving and serializing models](/guides/serialization_and_saving/).

## Privileged `training` argument in the `call()` method

Some layers, in particular the `BatchNormalization` layer and the `Dropout`
layer, have different behaviors during training and inference. For such
layers, it is standard practice to expose a `training` (boolean) argument in
the `call()` method.

By exposing this argument in `call()`, you enable the built-in training and
evaluation loops (e.g. `fit()`) to correctly use the layer in training and
inference.

```{r}
CustomDropout(keras$layers$Layer) %py_class% {
  initialize <- function(rate, ...) {
    super$initialize(...)
    self$rate <- rate
  }

>   call <- function(inputs, training = NULL) {

Reviewer: is NULL a good default? instead of FALSE i mean

"deep neural network").

So if you're wondering, "should I use the `Layer` class or the `Model` class?",
ask yourself: will I need to call `fit()` on it? Will I need to call `save()`
on it? If so, go with `Model`. If not (either because your class is just a block
in a bigger system, or because you are writing training & saving code yourself),
use `Layer`.

For instance, we could take our mini-resnet example above, and use it to build
a `Model` that we could train with `fit()`, and that we could save with
`save_model_weights_tf()`:

```{r, eval = FALSE}
ResNet(keras$Model) %py_class% {
  initialize <- function(num_classes = 1000) {
    super$initialize()
    self$block_1 <- ResNetBlock()
    self$block_2 <- ResNetBlock()
    self$global_pool <- layer_global_average_pooling_2d()
    self$classifier <- layer_dense(units = num_classes)
  }

  call <- function(inputs) {
    x <- self$block_1(inputs)
    x <- self$block_2(x)
    x <- self$global_pool(x)
    self$classifier(x)
  }
}

resnet <- ResNet()
dataset <- ...
resnet %>% fit(dataset, epochs = 10)
resnet %>% save_model_tf(filepath)
```

```

```
## Putting it all together: an end-to-end example

Here's what you've learned so far:

> - A `Layer` encapsulate a state (created in `initialize()` or `build()`) and some

Reviewer: ```suggestion
- A `Layer` encapsulates a state (created in `initialize()` or `build()`), and some
```

        ## with), When in doubt about the objects or methods that are available
        ## (e.g., what is `tape` here?), remember you can always drop into a
        ## debugger right here:
        # browser()

        reconstructed <- vae(x_batch_train)
        # Compute reconstruction loss
        loss <- mse_loss_fn(x_batch_train, reconstructed)

        loss %<>% add(vae$losses[[1]]) # Add KLD regularization loss
      })
      grads <- tape$gradient(loss, vae$trainable_weights)
      optimizer$apply_gradients(zip(grads, vae$trainable_weights))

      loss_metric(loss)

      step %<>% add(1)
      if (step %% 100 == 0) {
        cat(sprintf("step %d: mean loss = %.4f\n", step, loss_metric$result()))
      }
    }
  })
}
```

Note that since the VAE is subclassing `Model`, it features built-in training
loops. So you could also have trained it like this:

```{r}
vae <- VariationalAutoEncoder(784, 64, 32)

optimizer <- optimizer_adam(learning_rate = 1e-3)

vae %>% compile(optimizer, loss = loss_mean_squared_error())
vae %>% fit(x_train, x_train, epochs = 2, batch_size = 64)
```

## Beyond object-oriented development: the Functional API

> Was this example too much object-oriented development for you? You can also

Reviewer: ```suggestion
If you prefer a less object-oriented way of programming, you can also
```
